### ISSUE TITLE
Rule Change: Check that `cause` is passed to specified custom errors in `preserve-caught-error` rule

### ISSUE THREAD

**Issue by nathanhleung** [eslint/eslint#20428]
### What rule do you want to change?

 preserve-caught-error

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
class CustomApplicationError extends Error {}

try {
  doSomething();
} catch (err) {
  // With `"preserve-caught-error": "error", this generates an error:
  throw new Error("Cause not provided");

  // But a custom error that extends `Error` does not:
  throw new CustomApplicationError("Cause not provided, but no ESLint error!");
}
```

### What does the rule currently do for this code?

The rule properly marks the `throw new Error("Cause not provided");` line as failing the rule, but does not show any indication for the `throw new CustomApplicationError("Cause not provided, but no ESLint error!");` line.

### What will the rule do after it's changed?

A few possibilities:

1. ESLint can detect whether the custom error class inherits from the native `Error` (unsure if this possible in vanilla ESLint/without type information; I imagine this also could be brittle if constructor signature is changed in inheriting class?) and apply the rule to those custom error types as well.
2. The user can specify a list of custom error class names in the ESLint config to apply the rule to; ESLint would assume that these custom error classes accept a second parameter with a `cause` key in the same way native `Error`s do
3. Whenever an expression follows a `throw`, somehow check that the originally caught `err` is used somewhere in the expression?

### Participation

- [x] I am willing to submit a pull request to implement this change.

**Comment by DMartens** [comment]
Hi, thank you for wanting to contribute.
Your possibility number 2 was part of the initial design of the rule which was scrapped as this could be added later ([discussion](https://github.com/eslint/eslint/pull/19913#discussion_r2192608593)).
So we can use this issue to discuss the design of the option.
Some design questions would be:
- name of the option (your proposal used `errorClassNames` and `customErrorTypes` from the original PR)
- should regular expression patterns be supported (e.g. to check every error one could use "Error$")
- do we require that the variable is not declared locally (one of the raised concerns in the discussion; so either be imported or a global)
- we could also allow providing an object which allows specifying where the error variable is declared (e.g. `{ name: string, source: "import" | "global" | "any" }`)

An alternative (your possibility 1) would be to propose an extension rule to [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) as core rules are currently not type-aware.

**Comment by nathanhleung** [comment]
My thoughts:

> name of the option (your proposal used `errorClassNames` and `customErrorTypes` from the original PR)

I personally lean towards an option name containing `names` rather than `types` to mirror the [`.name`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/name) property on native `Error`s

> should regular expression patterns be supported (e.g. to check every error one could use "Error$")

Regular expression properties could be a nice extension, but a minimal version of this feature could support only strings to start.

> do we require that the variable is not declared locally (one of the raised concerns in the discussion; so either be imported or a global)

Was reading the [discussion](https://github.com/eslint/eslint/pull/19913#discussion_r2192608593) — why does the rule not report in this case even if `"SafeError"` is specified in the proposed option? cc @JoshuaKGoldberg 

```js
class SafeError { /* ... */ }

export function doSomethingDangerous() {
  // ...
  try {
    // ...
  } catch (error) {
    throw new SafeError(error);
  }
}
```

> we could also allow providing an object which allows specifying where the error variable is declared (e.g. { name: string, source: "import" | "global" | "any" })

This seems like a reasonable solution.

**Comment by fasttime** [comment]
Many error types don't support a `cause` property, and that's why the rule does not enforce supplying one when an arbitrary error (or other expression) is thrown from a catch block.

When `preserve-caught-error` was developed, we decided to defer the discussion about supporting custom error classes in order not to delay the introduction of the rule. That said, I agree that providing a mechanism for custom error support would be useful.

**Comment by snitin315** [comment]
I am in favor of this proposal. A new option like `errorClassNames` sounds good to me. Marking as accepted.

@nathanhleung Are you willing to submit a PR for this? Otherwise, I could take a look at this.

**Comment by Abilash-AvantGarde** [comment]
Hi everyone,
I’ve been following the discussion on this rule change and would love to contribute by implementing the errorClassNames option as discussed. Is this issue still available for me to take on? If so, I’m happy to start working on a Pull Request and will make sure to include the necessary tests.

Thank you!

cc @snitin315 @nathanhleung @DMartens @fasttime

**Comment by lumirlumir** [comment]
Hi @nathanhleung, are you still available to work on this issue?

If you don’t have time right now, that’s totally fine — happy to hand it off to someone else who volunteered.

If you are available, your PR would be welcome.

**Comment by nathanhleung** [comment]
> Hi @nathanhleung, are you still available to work on this issue?
> 
> If you don’t have time right now, that’s totally fine — happy to hand it off to someone else who volunteered.
> 
> If you are available, your PR would be welcome.

Hi, thanks for checking back in. Unfortunately not available right now so let's hand it off for now. Excited to see this move forward!

**Comment by Abilash-AvantGarde** [comment]
Hi @lumirlumir,

I’d be happy to help implement this feature.

I previously opened a PR with an `errorClassNames` option, but closed it as requested while waiting for consensus on this issue.

I’d be happy to reopen or update the PR accordingly.

**Comment by snitin315** [comment]
@Abilash-AvantGarde Feel free to send a PR

**Comment by meganprice661-svg** [comment]
Hi @Abilash-AvantGarde, thanks for the update! It looks like the issue is
being addressed in PR #20623, so I'll follow along with that. Appreciate
the initiative, and looking forward to seeing it move forward!
